### PR TITLE
AOTAutograd: keep input mutations in the graph if they are under no_grad, even if they require_grad

### PR DESCRIPTION
**Pull request by bdhirsh**
Quick recap of events:

(1) https://github.com/pytorch/pytorch/pull/111347, which fixed a perf regression in 2.1 compared to 2.0, introduced a correctness problem around input mutations on inputs that require grad that show up in an inference-only graph (the specific case where this can happen is rare and nobody reported the issue, but it was fixed a few weeks later)

(2) That fix happened here: https://github.com/pytorch/pytorch/pull/113584, which makes sure to keep input mutations outside of the graph, so the autograd engine can set metadata properly on them

(3) That in turn caused a slight regression compared to (1), which is what this PR attempts to fix. In particular, code like the below is safe to keep the mutations in the graph for:

```
@torch.compile
def f(x):
    x.mul_(2)
    
x = torch.ones(2, requires_grad=True).clone()
# x requires_grad, so the input mutation will change some autograd metadata, like the version counter
# However, the mutation is under no_grad, so we don't have to worry about e.g. aliases of x having their .grad_fn fields changed
with torch.no_grad():
    f(x)
```

This particular case is pretty important to the shampoo optimizer code, which is run under `torch.compile`, and mutates parameters (which require grad).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113417
* __->__ #114646

### CODE REVIEW

File: torch/csrc/autograd/python_torch_functions_manual.cpp [hunk 1]
```diff
@@ -663,6 +663,28 @@ THPVariable__functionalize_are_all_mutations_hidden_from_autograd(
...
 }
 
+static PyObject*
+THPVariable__functionalize_are_all_mutations_under_no_grad_or_inference_mode(
+    PyObject* self,
+    PyObject* args,
+    PyObject* kwargs) {
+  HANDLE_TH_ERRORS
+  static PythonArgParser parser(
+      {"_functionalize_are_all_mutations_under_no_grad_or_inference_mode(Tensor t)"},
```
Comment (zou3519) [inline torch/csrc/autograd/python_torch_functions_manual.cpp:673]: It's easier to pybind these in e.g. https://github.com/pytorch/pytorch/blob/main/torch/csrc/utils/python_dispatch.cpp#L738 than to manually write bindings like this.

**Review by zou3519 — APPROVED** [review]

**Comment by bdhirsh** [comment]
Some of the triton kernel tests are failing aot_autograd asserts in CI - looking into them

File: torch/csrc/autograd/python_torch_functions_manual.cpp (hunk 1 above)
Comment (bdhirsh) [reply torch/csrc/autograd/python_torch_functions_manual.cpp:673]: you're right - at some point I should move all of these functionalization bindings to pybind

**Pull request closed without merging**